### ISSUE TITLE
Cannot run code on A6000 GPU

### ISSUE THREAD

**Issue by subhalingamd** [stanford-futuredata/ColBERT#117]
Hi!

I am not being able to run indexing with *NVIDIA RTX A6000* GPU (uses *compute_86*). I get an `Unsupported gpu architecture 'compute_86'` error while the code tries to compile `colbert/indexing/codecs/decompress_residuals.cu`.

<details>
<summary> Complete traceback </summary>

```
Traceback (most recent call last):
  File "/home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 1814, in _run_ninja_build
    env=env)
  File "/home/subhalingam/miniconda3/envs/colbert/lib/python3.7/subprocess.py", line 512, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['ninja', '-v']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/subhalingam/colbert/ColBERT/colbert/__init__.py", line 2, in <module>
    from .indexer import Indexer
  File "/home/subhalingam/colbert/ColBERT/colbert/indexer.py", line 12, in <module>
    from colbert.indexing.collection_indexer import encode
  File "/home/subhalingam/colbert/ColBERT/colbert/indexing/collection_indexer.py", line 24, in <module>
    from colbert.indexing.index_saver import IndexSaver
  File "/home/subhalingam/colbert/ColBERT/colbert/indexing/index_saver.py", line 8, in <module>
    from colbert.indexing.codecs.residual import ResidualCodec
  File "/home/subhalingam/colbert/ColBERT/colbert/indexing/codecs/residual.py", line 30, in <module>
    verbose=os.getenv("COLBERT_LOAD_TORCH_EXTENSION_VERBOSE", "False") == "True",
  File "/home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 1214, in load
    keep_intermediates=keep_intermediates)
  File "/home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 1435, in _jit_compile
    is_standalone=is_standalone)
  File "/home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 1540, in _write_ninja_file_and_build_library
    error_prefix=f"Error building extension '{name}'")
  File "/home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/utils/cpp_extension.py", line 1824, in _run_ninja_build
    raise RuntimeError(message) from e
RuntimeError: Error building extension 'decompress_residuals_cpp': [1/2] /usr/bin/nvcc  -DTORCH_EXTENSION_NAME=decompress_residuals_cpp -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1013\" -isystem /home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/include -isystem /home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/include/torch/csrc/api/include -isystem /home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/include/TH -isystem /home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/include/THC -isystem /home/subhalingam/miniconda3/envs/colbert/include/python3.7m -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_86,code=compute_86 -gencode=arch=compute_86,code=sm_86 --compiler-options '-fPIC' -std=c++14 -c /home/subhalingam/colbert/ColBERT/colbert/indexing/codecs/decompress_residuals.cu -o decompress_residuals.cuda.o 
FAILED: decompress_residuals.cuda.o 
/usr/bin/nvcc  -DTORCH_EXTENSION_NAME=decompress_residuals_cpp -DTORCH_API_INCLUDE_EXTENSION_H -DPYBIND11_COMPILER_TYPE=\"_gcc\" -DPYBIND11_STDLIB=\"_libstdcpp\" -DPYBIND11_BUILD_ABI=\"_cxxabi1013\" -isystem /home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/include -isystem /home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/include/torch/csrc/api/include -isystem /home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/include/TH -isystem /home/subhalingam/miniconda3/envs/colbert/lib/python3.7/site-packages/torch/include/THC -isystem /home/subhalingam/miniconda3/envs/colbert/include/python3.7m -D_GLIBCXX_USE_CXX11_ABI=0 -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr -gencode=arch=compute_86,code=compute_86 -gencode=arch=compute_86,code=sm_86 --compiler-options '-fPIC' -std=c++14 -c /home/subhalingam/colbert/ColBERT/colbert/indexing/codecs/decompress_residuals.cu -o decompress_residuals.cuda.o 
nvcc fatal   : Unsupported gpu architecture 'compute_86'
ninja: build stopped: subcommand failed.
```

</details>

The above error can be reproduced by simply running the following line on python shell:
```python
from colbert.infra import Run, RunConfig, ColBERTConfig
```

In `conda_env.yml`, I changed the `cudatoolkit` version to 11.3 and additionally included `cupy-cuda113` under the `pip` list. I also removed the version number requirements in couple of packages (but I do not think they are crucial).

<details>
<summary> Updated conda_env.yml </summary>

```
name: colbert
channels:
  - pytorch
  - nvidia
  - conda-forge
  - defaults
dependencies:
  - python=3.7
  - faiss-gpu=1.7.0
  - pip=21.0
  - pytorch
  - torchaudio
  - torchvision
  - cudatoolkit=11.3
  - pip:
    - cupy-cuda113
    - bitarray
    - gitpython
    - jupyter
    - jupyterlab
    - ninja
    - scipy
    - spacy
    - tqdm
    - transformers==4.10.0
    - ujson
```

</details>

<details>
<summary> Conda environment info </summary>

```
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                 conda_forge    conda-forge
_openmp_mutex             4.5                  2_kmp_llvm    conda-forge
anyio                     3.6.1                    pypi_0    pypi
argon2-cffi               21.3.0                   pypi_0    pypi
argon2-cffi-bindings      21.2.0                   pypi_0    pypi
attrs                     21.4.0                   pypi_0    pypi
babel                     2.10.3                   pypi_0    pypi
backcall                  0.2.0                    pypi_0    pypi
beautifulsoup4            4.11.1                   pypi_0    pypi
bitarray                  2.5.1                    pypi_0    pypi
blas                      2.115                       mkl    conda-forge
blas-devel                3.9.0            15_linux64_mkl    conda-forge
bleach                    5.0.1                    pypi_0    pypi
blis                      0.7.8                    pypi_0    pypi
brotlipy                  0.7.0           py37h540881e_1004    conda-forge
bzip2                     1.0.8                h7f98852_4    conda-forge
ca-certificates           2022.6.15            ha878542_0    conda-forge
catalogue                 2.0.7                    pypi_0    pypi
certifi                   2022.6.15        py37h89c1867_0    conda-forge
cffi                      1.15.1           py37h43b0acd_0    conda-forge
charset-normalizer        2.1.0              pyhd8ed1ab_0    conda-forge
click                     8.1.3                    pypi_0    pypi
cryptography              37.0.4           py37h38fbfac_0    conda-forge
cudatoolkit               11.3.1               ha36c431_9    nvidia
cupy-cuda113              10.6.0                   pypi_0    pypi
cymem                     2.0.6                    pypi_0    pypi
debugpy                   1.6.2                    pypi_0    pypi
decorator                 5.1.1                    pypi_0    pypi
defusedxml                0.7.1                    pypi_0    pypi
entrypoints               0.4                      pypi_0    pypi
faiss                     1.7.0           py37cuda112h7ce9dcd_8_cuda    conda-forge
faiss-gpu                 1.7.0                h788eb59_8    conda-forge
fastjsonschema            2.15.3                   pypi_0    pypi
fastrlock                 0.8                      pypi_0    pypi
ffmpeg                    4.3                  hf484d3e_0    pytorch
filelock                  3.7.1                    pypi_0    pypi
freetype                  2.10.4               h0708190_1    conda-forge
giflib                    5.2.1                h36c2ea0_2    conda-forge
gitdb                     4.0.9                    pypi_0    pypi
gitpython                 3.1.27                   pypi_0    pypi
gmp                       6.2.1                h58526e2_0    conda-forge
gnutls                    3.6.13               h85f3911_1    conda-forge
huggingface-hub           0.8.1                    pypi_0    pypi
idna                      3.3                pyhd8ed1ab_0    conda-forge
importlib-metadata        4.12.0                   pypi_0    pypi
importlib-resources       5.8.0                    pypi_0    pypi
ipykernel                 6.15.1                   pypi_0    pypi
ipython                   7.34.0                   pypi_0    pypi
ipython-genutils          0.2.0                    pypi_0    pypi
ipywidgets                7.7.1                    pypi_0    pypi
jedi                      0.18.1                   pypi_0    pypi
jinja2                    3.1.2                    pypi_0    pypi
joblib                    1.1.0                    pypi_0    pypi
jpeg                      9e                   h166bdaf_2    conda-forge
json5                     0.9.8                    pypi_0    pypi
jsonschema                4.7.2                    pypi_0    pypi
jupyter                   1.0.0                    pypi_0    pypi
jupyter-client            7.3.4                    pypi_0    pypi
jupyter-console           6.4.4                    pypi_0    pypi
jupyter-core              4.11.1                   pypi_0    pypi
jupyter-server            1.18.1                   pypi_0    pypi
jupyterlab                3.4.3                    pypi_0    pypi
jupyterlab-pygments       0.2.2                    pypi_0    pypi
jupyterlab-server         2.15.0                   pypi_0    pypi
jupyterlab-widgets        1.1.1                    pypi_0    pypi
lame                      3.100             h7f98852_1001    conda-forge
langcodes                 3.3.0                    pypi_0    pypi
lcms2                     2.12                 hddcbb42_0    conda-forge
ld_impl_linux-64          2.36.1               hea4e1c9_2    conda-forge
lerc                      3.0                  h9c3ff4c_0    conda-forge
libblas                   3.9.0            15_linux64_mkl    conda-forge
libcblas                  3.9.0            15_linux64_mkl    conda-forge
libdeflate                1.12                 h166bdaf_0    conda-forge
libfaiss                  1.7.0           cuda112h5bea7ad_8_cuda    conda-forge
libfaiss-avx2             1.7.0           cuda112h1234567_8_cuda    conda-forge
libffi                    3.4.2                h7f98852_5    conda-forge
libgcc-ng                 12.1.0              h8d9b700_16    conda-forge
libgfortran-ng            12.1.0              h69a702a_16    conda-forge
libgfortran5              12.1.0              hdcd56e2_16    conda-forge
libiconv                  1.17                 h166bdaf_0    conda-forge
liblapack                 3.9.0            15_linux64_mkl    conda-forge
liblapacke                3.9.0            15_linux64_mkl    conda-forge
libnsl                    2.0.0                h7f98852_0    conda-forge
libpng                    1.6.37               h753d276_3    conda-forge
libstdcxx-ng              12.1.0              ha89aaad_16    conda-forge
libtiff                   4.4.0                hc85c160_1    conda-forge
libwebp                   1.2.2                h3452ae3_0    conda-forge
libwebp-base              1.2.2                h7f98852_1    conda-forge
libxcb                    1.13              h7f98852_1004    conda-forge
libzlib                   1.2.12               h166bdaf_1    conda-forge
llvm-openmp               14.0.4               he0ac6c6_0    conda-forge
lz4-c                     1.9.3                h9c3ff4c_1    conda-forge
markupsafe                2.1.1                    pypi_0    pypi
matplotlib-inline         0.1.3                    pypi_0    pypi
mistune                   0.8.4                    pypi_0    pypi
mkl                       2022.1.0           h84fe81f_915    conda-forge
mkl-devel                 2022.1.0           ha770c72_916    conda-forge
mkl-include               2022.1.0           h84fe81f_915    conda-forge
murmurhash                1.0.7                    pypi_0    pypi
nbclassic                 0.4.3                    pypi_0    pypi
nbclient                  0.6.6                    pypi_0    pypi
nbconvert                 6.5.0                    pypi_0    pypi
nbformat                  5.4.0                    pypi_0    pypi
ncurses                   6.3                  h27087fc_1    conda-forge
nest-asyncio              1.5.5                    pypi_0    pypi
nettle                    3.6                  he412f7d_0    conda-forge
ninja                     1.10.2.3                 pypi_0    pypi
notebook                  6.4.12                   pypi_0    pypi
notebook-shim             0.1.0                    pypi_0    pypi
numpy                     1.21.6           py37h976b520_0    conda-forge
openh264                  2.1.1                h780b84a_0    conda-forge
openjpeg                  2.4.0                hb52868f_1    conda-forge
openssl                   1.1.1q               h166bdaf_0    conda-forge
packaging                 21.3                     pypi_0    pypi
pandocfilters             1.5.0                    pypi_0    pypi
parso                     0.8.3                    pypi_0    pypi
pathy                     0.6.2                    pypi_0    pypi
pexpect                   4.8.0                    pypi_0    pypi
pickleshare               0.7.5                    pypi_0    pypi
pillow                    9.2.0            py37h44f0d7a_0    conda-forge
pip                       21.0.1             pyhd8ed1ab_0    conda-forge
preshed                   3.0.6                    pypi_0    pypi
prometheus-client         0.14.1                   pypi_0    pypi
prompt-toolkit            3.0.30                   pypi_0    pypi
psutil                    5.9.1                    pypi_0    pypi
pthread-stubs             0.4               h36c2ea0_1001    conda-forge
ptyprocess                0.7.0                    pypi_0    pypi
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pydantic                  1.9.1                    pypi_0    pypi
pygments                  2.12.0                   pypi_0    pypi
pyopenssl                 22.0.0             pyhd8ed1ab_0    conda-forge
pyparsing                 3.0.9                    pypi_0    pypi
pyrsistent                0.18.1                   pypi_0    pypi
pysocks                   1.7.1            py37h89c1867_5    conda-forge
python                    3.7.12          hb7a2778_100_cpython    conda-forge
python-dateutil           2.8.2                    pypi_0    pypi
python_abi                3.7                     2_cp37m    conda-forge
pytorch                   1.12.0          py3.7_cuda11.3_cudnn8.3.2_0    pytorch
pytorch-mutex             1.0                        cuda    pytorch
pytz                      2022.1                   pypi_0    pypi
pyyaml                    6.0                      pypi_0    pypi
pyzmq                     23.2.0                   pypi_0    pypi
qtconsole                 5.3.1                    pypi_0    pypi
qtpy                      2.1.0                    pypi_0    pypi
readline                  8.1.2                h0f457ee_0    conda-forge
regex                     2022.7.9                 pypi_0    pypi
requests                  2.28.1             pyhd8ed1ab_0    conda-forge
sacremoses                0.0.53                   pypi_0    pypi
scipy                     1.7.3                    pypi_0    pypi
send2trash                1.8.0                    pypi_0    pypi
setuptools                63.1.0           py37h89c1867_0    conda-forge
six                       1.16.0                   pypi_0    pypi
smart-open                5.2.1                    pypi_0    pypi
smmap                     5.0.0                    pypi_0    pypi
sniffio                   1.2.0                    pypi_0    pypi
soupsieve                 2.3.2.post1              pypi_0    pypi
spacy                     3.4.0                    pypi_0    pypi
spacy-legacy              3.0.9                    pypi_0    pypi
spacy-loggers             1.0.2                    pypi_0    pypi
sqlite                    3.39.0               h4ff8645_0    conda-forge
srsly                     2.4.3                    pypi_0    pypi
tbb                       2021.5.0             h924138e_1    conda-forge
terminado                 0.15.0                   pypi_0    pypi
thinc                     8.1.0                    pypi_0    pypi
tinycss2                  1.1.1                    pypi_0    pypi
tk                        8.6.12               h27826a3_0    conda-forge
tokenizers                0.10.3                   pypi_0    pypi
torchaudio                0.12.0               py37_cu113    pytorch
torchvision               0.13.0               py37_cu113    pytorch
tornado                   6.2                      pypi_0    pypi
tqdm                      4.64.0                   pypi_0    pypi
traitlets                 5.3.0                    pypi_0    pypi
transformers              4.10.0                   pypi_0    pypi
typer                     0.4.2                    pypi_0    pypi
typing-extensions         4.1.1                    pypi_0    pypi
ujson                     5.4.0                    pypi_0    pypi
urllib3                   1.26.10            pyhd8ed1ab_0    conda-forge
wasabi                    0.9.1                    pypi_0    pypi
wcwidth                   0.2.5                    pypi_0    pypi
webencodings              0.5.1                    pypi_0    pypi
websocket-client          1.3.3                    pypi_0    pypi
wheel                     0.37.1             pyhd8ed1ab_0    conda-forge
widgetsnbextension        3.6.1                    pypi_0    pypi
xorg-libxau               1.0.9                h7f98852_0    conda-forge
xorg-libxdmcp             1.1.3                h7f98852_0    conda-forge
xz                        5.2.5                h516909a_1    conda-forge
zipp                      3.8.0                    pypi_0    pypi
zlib                      1.2.12               h166bdaf_1    conda-forge
zstd                      1.5.2                h8a70e8d_2    conda-forge
```

</details>

Any help in this regard would be greatly appreciated. 

Thanks.

**Comment by santhnm2** [comment]
Hi @subhalingamd, just to confirm, are you using the `main` branch? And if so could you try prepending your command with `COLBERT_LOAD_TORCH_EXTENSION_VERSBOSE=True` and paste the output? I suspect that `nvcc`is trying to use the system CUDA installation (as opposed to the conda installation) which is perhaps out of date for Ampere architectures.

Also, if you're using the latest `main` branch you shouldn't need the `cupy` dependency anymore - please let me know if run into any `cupy` issues, this would definitely be unexpected behavior.

**Comment by subhalingamd** [comment]
Hi @santhnm2,

Thanks for pointing out at the possible `nvcc` version mismatch. It was indeed the cause for the error- the system had an older version for some reason. I downloaded `cudatoolkit-dev=11.3` (instead of `cudatoolkit=11.3`) which also installed the correct version of `nvcc` in the environment. This fixed the issue.

And yes, I am using the `main` branch. Sorry for not mentioning that explicitly.

Regarding the `cupy` dependency, I cloned the repo before the commit 1f0ece2 which required it. 

Thanks a lot for a quick response!
